src/simulators/statevector/qubitvector.hpp

```diff
@@ -1801,150 +1709,102 @@ void QubitVector<data_t>::apply_diagonal_matrix(const uint_t qubit,
-
-    reg_t sorted_src_qubits = src_qubits;
-    std::sort(sorted_src_qubits.begin(), sorted_src_qubits.end());
-    const cvector_t sorted_vmat = sort_matrix(src_qubits, sorted_src_qubits, vmat);
-
-    // 1. identify low and high delta
-    auto low = std::distance(dst_sorted_qubits.begin(),
-                               std::find(dst_sorted_qubits.begin(), dst_sorted_qubits.end(), sorted_src_qubits[0]));
-
-    auto high = std::distance(dst_sorted_qubits.begin(),
-                                std::find(dst_sorted_qubits.begin(), dst_sorted_qubits.end(), sorted_src_qubits[1]));
-
-    auto low_delta = BITS[low];
-    auto high_delta = BITS[high];
-
-    // 2. find op.u(0, 0) position in U
-    for (uint_t i = 0; i < dst_dim; ++i) {
-      if (filled[i])
-        continue;
-
-      //  3. allocate vmat to u based on u(i, i) and delta
-      u[i                             + (i + 0)                      * dst_dim] = sorted_vmat[0 + 0 * src_dim];
-      u[i                             + (i + low_delta)              * dst_dim] = sorted_vmat[0 + 1 * src_dim];
-      u[i                             + (i + high_delta)             * dst_dim] = sorted_vmat[0 + 2 * src_dim];
-      u[i                             + (i + low_delta + high_delta) * dst_dim] = sorted_vmat[0 + 3 * src_dim];
-      u[(i + low_delta)               + (i + 0)                      * dst_dim] = sorted_vmat[1 + 0 * src_dim];
-      u[(i + low_delta)               + (i + low_delta)              * dst_dim] = sorted_vmat[1 + 1 * src_dim];
-      u[(i + low_delta)               + (i + high_delta)             * dst_dim] = sorted_vmat[1 + 2 * src_dim];
-      u[(i + low_delta)               + (i + low_delta + high_delta) * dst_dim] = sorted_vmat[1 + 3 * src_dim];
-      u[(i + high_delta)              + (i + 0)                      * dst_dim] = sorted_vmat[2 + 0 * src_dim];
-      u[(i + high_delta)              + (i + low_delta)              * dst_dim] = sorted_vmat[2 + 1 * src_dim];
-      u[(i + high_delta)              + (i + high_delta)             * dst_dim] = sorted_vmat[2 + 2 * src_dim];
-      u[(i + high_delta)              + (i + low_delta + high_delta) * dst_dim] = sorted_vmat[2 + 3 * src_dim];
-      u[(i + low_delta + high_delta)  + (i + 0)                      * dst_dim] = sorted_vmat[3 + 0 * src_dim];
-      u[(i + low_delta + high_delta)  + (i + low_delta)              * dst_dim] = sorted_vmat[3 + 1 * src_dim];
-      u[(i + low_delta + high_delta)  + (i + high_delta)             * dst_dim] = sorted_vmat[3 + 2 * src_dim];
-      u[(i + low_delta + high_delta)  + (i + low_delta + high_delta) * dst_dim] = sorted_vmat[3 + 3 * src_dim];
-
-      filled[i] = true;
-      filled[i + low_delta] = true;
-      filled[i + high_delta] = true;
-      filled[i + low_delta + high_delta] = true;
-    }
-    //TODO: } else if (src_qubits.size() == 3) {
+  if (mat[0] == 1.0 && mat[4] == 0.0 && mat[8]  == 0.0 && mat[12] == 0.0
```

Comment:
Ok, this is very tought to read.
Can we make a helper function/lambda that returns `true` or `false` to add some meaning to this conditional?
Something like, I don't know:
```c++
if( isWhateverMeangfullMatrixFormName(mat) ){ 
```

Reply:
Hummm, a bit difficult to express these kinds of matrixes...

Reply:
I removed specialization for 2qubit matrix operation. diagonal matrix operation covers all.